cpp/lib/src/FileLocker.cc

```diff
@@ -34,22 +35,36 @@
 #include "util.h"
 
 
-FileLocker::FileLocker(const int fd, const LockType lock_type)
-    : lock_fd_(fd)
+FileLocker::FileLocker(const int fd, const LockType lock_type, const unsigned timeout)
+    : lock_fd_(fd), interrupted_(false)
 {
+    if (timeout > 0) {
+        if (::alarm(timeout) != 0)
+            LOG_ERROR("we already had an active alarm!");
+    }
+
     struct flock lock_struct;
     lock_struct.l_type   = static_cast<short>((lock_type == READ_ONLY) ? F_RDLCK : F_WRLCK); /* F_RDLCK, F_WRLCK, F_UNLCK    */
     lock_struct.l_whence = SEEK_SET;                                                         /* SEEK_SET, SEEK_CUR, SEEK_END */
     lock_struct.l_start  = 0;                                                                /* Offset from l_whence         */
     lock_struct.l_len    = 0;                                                                /* length, 0 = to EOF           */
     lock_struct.l_pid    = ::getpid();                                                       /* our PID                      */
 
-    if (::fcntl(lock_fd_, F_SETLKW, &lock_struct) == -1)
-        throw std::runtime_error("in FileLocker::FileLocker: fcntl(2) failed! (" + std::to_string(errno) + ")!");
+    if (::fcntl(lock_fd_, F_SETLKW, &lock_struct) == -1) {
+        const int last_errno(errno);
+        ::alarm(0); // Cancel our alarm!
+        if (last_errno != EINTR)
+            throw std::runtime_error("fcntl(2) failed! (" + std::to_string(last_errno) + ")!");
+        errno = 0;
+        interrupted_ = true;
```

Comment:
So this will be used later to prevent an unlock.

But if we get here this also means that the locking has never been complete, so how do we guarantee a file is locked after this function ends? Shouldn't it be called in a loop while the locking hasn't been successful?